### PR TITLE
Fixes the issue - Website appears right-to-left ish on disabling cookies

### PR DESCRIPTION
**Pull request by share-with-me**
This fix lets the JS script interact with the localStorage of the browser only when the cookies are enabled. Solves the issue #99 . A demo can be found [here](https://drive.google.com/open?id=0BwrSYlOsMCyWWUFzN2Nlci1QcXM)

### CODE REVIEW

**Comment by sushain97** [comment]
What is the root cause? Why does having cookies disabled break the interface?

**Comment by share-with-me** [comment]
@sushain97 Unable to find the root cause. My approach was, let the code interact with `store.js` and cache contents only when the cookies are enabled. I experimented, put a condition of checking if the cookies are enabled and it seemed to work! The site worked perfectly in the case when even if the user disables the cookies. It seems a solution is at hand but am still trying to get to the root cause.

**Comment by sushain97** [comment]
> The site worked perfectly in the case when even if the user disables the cookies.

Unfortunately, that's impossible to determine since we don't have unit/integration tests :)

> It seems a solution is at hand but am still trying to get to the root cause.

Hm, okay. That's a good start but before merging any solution I would like to know why it fixes the problem. This way we can be certain that there are no inadvertent regressions.

**Comment by share-with-me** [comment]
> Unfortunately, that's impossible to determine since we don't have unit/integration tests :)

Yes, indeed! My bad! I meant it did not break interface appearance wise! Yes, unit/integration tests are needed.

> Hm, okay. That's a good start but before merging any solution I would like to know why it fixes the problem. This way we can be certain that there are no inadvertent regressions.

Yes, I would keep trying to get to the root cause of it.
Thanks :D.

**Comment by share-with-me** [comment]
After digging up a bit, this [stackoverflow link](https://stackoverflow.com/questions/26550770/html5-can-session-storage-local-storage-be-disabled-and-cookies-enabled), discussed about how localStorage works for different browsers in case the cookies are disabled. It mentions that for Chrome and FF, when we disable the cookies, the DOM Storage is automatically shut. However, for IE, it does remain open. 

The error mentioned for FF (in the stackoverflow link) is similar to the one I get in the console as in the following screenshot.
Does that mean that when we disable the cookies and try `Store.prototype.*` operations, the browsers raise an error and stop the page from loading further?!

I did not understand what `if(this.able()) { }` does in the `store.js` code. 

I believe the root cause is interacting with the `localStorage` of browsers in the event of cookies being disabled.

Please find the screenshot below.
screen shot 2017-06-06 at 9 53 16 am

**Comment by sushain97** [comment]
> Does that mean that when we disable the cookies and try Store.prototype.* operations, the browsers raise an error and stop the page from loading further?!

Perhaps. What's on min.js L434?

> I did not understand what if(this.able()) { } does in the store.js code.

https://github.com/goavki/apertium-html-tools/blob/master/assets/js/store.js#L63-L65

checks whether LocalStorage is usable.

> I believe the root cause is interacting with the localStorage of browsers in the event of cookies being disabled.

If this is true, the proper solution is patching `Store.able` to me more accurate. It should be a < 5 line change.

**Comment by unhammer** [comment]
> If this is true (which I agree with being the case on all but IE 11), the proper solution is patching Store.able to me more accurate. It should be a < 5 line change. However, since IE 11 cookies disabled != LocalStorage disabled, the ideal solution requires finding out how to figure out that LocalStorage is disabled reliably since the current Store.able implementation seems deficient.

If I understand correctly, I think we should avoid special-casing IE11 here. Just say that *Store requires both cookies and localstorage* to be `able`. (If Firefox and Chromium don't support localStorage when cookies are off, we shouldn't write special code to make IE do that).

**Comment by unhammer** [comment]
> If this is true (which I agree with being the case on all but IE 11), the proper solution is patching Store.able to me more accurate. It should be a < 5 line change. However, since IE 11 cookies disabled != LocalStorage disabled, the ideal solution requires finding out how to figure out that LocalStorage is disabled reliably since the current Store.able implementation seems deficient.

If I understand correctly, I think we should avoid special-casing IE11 here. Just say that *Store requires both cookies and localstorage* to be `able`. This should require no changes to `persistence.js` or any other file than `store.js`. (If Firefox and Chromium don't support localStorage when cookies are off, we shouldn't write special code to make IE do that).

**Comment by sushain97** [comment]
> If I understand correctly, I think we should avoid special-casing IE11 here. Just say that Store requires both cookies and localstorage to be able.

Hm, what if FF/Chrome changes the behavior of that checkbox to at some point not also disable LocalStorage? Would it not be more future-proof to reliably detect whether LocalStorage is available (which incidentally will make it work on IE11)?

**Comment by share-with-me** [comment]
> If this is true (which I agree with being the case on all but IE 11), the proper solution is patching Store.able to me more accurate. It should be a < 5 line change. However, since IE 11 cookies disabled != LocalStorage disabled, the ideal solution requires finding out how to figure out that LocalStorage is disabled reliably since the current Store.able implementation seems deficient.

@sushain97  So if I am not wrong, we need to make `Store.able()` efficient enough to understand if the `localStorage` is accessible. Thus, if someone disabled the cookies, `localsStorage` automatically may shut off ( FF/Chrome ) and `Store.able()` will be able to find that out and prevent all the code inside `if( Store.able() ) { }`  from executing!

**Comment by share-with-me** [comment]
@sushain97 , Please have a look! A try-catch logic to handle the same.

File: assets/js/store.js [hunk 1]
```diff
@@ -61,7 +61,13 @@ Store.prototype.has = function (key/*: string*/)/*: bool*/ {
 };
 
 Store.prototype.able = function ()/*: bool*/ {
-    return !!(window.localStorage);
+    try {
+        return !!(window.localStorage);
+    }
+    catch(e) {
+        console.warn('Cookies are disabled. Thus, unable to access localStorage.');
```
Comment (sushain97) [inline assets/js/store.js]: Is it possible to only handle the SecurityException? e.g. something like https://stackoverflow.com/questions/1433558/handling-specific-errors-in-javascript-think-exceptions

Comment (share-with-me) [inline assets/js/store.js]: Yes, makes sense! On it.

**Comment by share-with-me** [comment]
@sushain97 , Made the changes! Please review.

File: assets/js/store.js [hunk 2]
```diff
@@ -62,12 +62,26 @@ Store.prototype.has = function (key/*: string*/)/*: bool*/ {
 
 Store.prototype.able = function ()/*: bool*/ {
     try {
+        throwLocalStorageNotAccessibleError();
```
Comment (sushain97) [inline assets/js/store.js]: Doesn't this always throw the exception? I meant catching the `SecurityException` that you showed Firefox throws, not creating your own exception.

Comment (share-with-me) [inline assets/js/store.js]: > Doesn't this always throw the exception? 

Yep. I tested it incorrectly. My bad!

>  I meant catching the SecurityException that you showed Firefox throws, not creating your own exception.

I will look for the error codes that exist for such exceptions! 
Thanks :D

Comment (share-with-me) [inline assets/js/store.js]: We are looking at something like this right?!

```
    try {
        return !!(window.localStorage);
    }
    catch(e) {
        if(e.name === 'SecurityError') { // for FF, do similar for chrome too
            console.warn('Cookies disabled. Unable to access local browser storage.');
            return false;
        }
        else {
            throw e;
        }
    }

```

Comment (sushain97) [inline assets/js/store.js]: Yes.

Comment (share-with-me) [inline assets/js/store.js]: Perfect! It seems `SecurityError` is the name of the exception in FF. Looking for something similar in Chrome!

**Comment by share-with-me** [comment]
@sushain97 Thank you so much :D. One more down :D

**Pull request merged**